# Search Engine implementation

Workshop as an introduction to the LLMZoomcamp course where a RAG ill be implemented. This is a system that allows to "chat" with your own data (See previous workshop). In the previous workshop we used elastic search to return information to be used as context for our LLM. This time we will talk about search, which is an important part of RAGs.

The idea is understanding the principles of a searchengine, but not to produce a produciton ready system.

In [1]:
# Download the data
import requests

docs_url = 'https://github.com/alexeygrigorev/llm-rag-workshop/raw/main/notebooks/documents.json'
docs_response = requests.get(docs_url)
documents_raw = docs_response.json()

documents = []

for course in documents_raw:
    course_name = course['course']

    for doc in course['documents']:
        doc['course'] = course_name
        documents.append(doc)

In [2]:
documents[0]

{'text': "The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
 'section': 'General course-related questions',
 'question': 'Course - When will the course start?',
 'course': 'data-engineering-zoomcamp'}

The idea is to build a search egine which searches over the fields (question, section and text in this case) and returns the most relevant documents 

In [3]:
# Create a Data frame
import pandas as pd

df = pd.DataFrame(
    documents, 
    columns=["course", "section", "question", "text"]
)

In [4]:
df.head()

course                           section  \
0  data-engineering-zoomcamp  General course-related questions   
1  data-engineering-zoomcamp  General course-related questions   
2  data-engineering-zoomcamp  General course-related questions   
3  data-engineering-zoomcamp  General course-related questions   
4  data-engineering-zoomcamp  General course-related questions   

                                            question  \
0               Course - When will the course start?   
1  Course - What are the prerequisites for this c...   
2  Course - Can I still join the course after the...   
3  Course - I have registered for the Data Engine...   
4   Course - What can I do before the course starts?   

                                                text  
0  The purpose of this document is to capture fre...  
1  GitHub - DataTalksClub data-engineering-zoomca...  
2  Yes, even if you don't register, you're still ...  
3  You don't need it. You're accepted. You can al...  
4  You can start by installing and setting up all...

In [5]:
df[df.course == 'data-engineering-zoomcamp']

course                           section  \
0    data-engineering-zoomcamp  General course-related questions   
1    data-engineering-zoomcamp  General course-related questions   
2    data-engineering-zoomcamp  General course-related questions   
3    data-engineering-zoomcamp  General course-related questions   
4    data-engineering-zoomcamp  General course-related questions   
..                         ...                               ...   
430  data-engineering-zoomcamp           Workshop 2 - RisingWave   
431  data-engineering-zoomcamp           Workshop 2 - RisingWave   
432  data-engineering-zoomcamp           Workshop 2 - RisingWave   
433  data-engineering-zoomcamp          Triggers in Mage via CLI   
434  data-engineering-zoomcamp          Triggers in Mage via CLI   

                                              question  \
0                 Course - When will the course start?   
1    Course - What are the prerequisites for this c...   
2    Course - Can I still join the course after the...   
3    Course - I have registered for the Data Engine...   
4     Course - What can I do before the course starts?   
..                                                 ...   
430  Unable to Open Dashboard as xdg-open doesn’t o...   
431  Resolving Python Interpreter Path Inconsistenc...   
432                    How does windowing work in Sql?   
433  Encountering the error "ModuleNotFoundError: N...   
434                                     Basic Commands   

                                                  text  
0    The purpose of this document is to capture fre...  
1    GitHub - DataTalksClub data-engineering-zoomca...  
2    Yes, even if you don't register, you're still ...  
3    You don't need it. You're accepted. You can al...  
4    You can start by installing and setting up all...  
..                                                 ...  
430  Refer to the solution given in the first solut...  
431  Example Error:\nWhen attempting to execute a P...  
432  Ans : Windowing in streaming SQL involves defi...  
433  Python 3.12.1, is not compatible with kafka-py...  
434  Docker Commands\n# Create a Docker Image from ...  

[435 rows x 4 columns]

we will usec **Vector spaces**, which means we turn the documents into vectors

* term-document matrix:
    * rows: documents
    * columns: words/tokens

* bag of words:
    * word order is lost
    * sparse matrix

In [6]:
from sklearn.feature_extraction.text import CountVectorizer

In [7]:
cv = CountVectorizer()

In [8]:
cv.fit(df.text)

CountVectorizer()

In [9]:
cv.get_feature_names_out()

array(['00', '00000000e', '0002', ..., '要了解键盘快捷键', '要启用屏幕阅读器支持', '请按ctrl'],
      dtype=object)

In [10]:
cv.get_feature_names_out().shape

(6711,)

This means that in the "text" column of the dataset we have 6711 different tokens, that seems to include a lot of noise. Therefore we may only care for the tokens that appear at least in a certain number of documents.

In [11]:
cv = CountVectorizer(min_df=5)
cv.fit(df.text)

CountVectorizer(min_df=5)

In [12]:
cv.get_feature_names_out()

array(['01', '02', '03', ..., 'youtube', 'zip', 'zoomcamp'], dtype=object)

In [13]:
cv.get_feature_names_out().shape

(1524,)

In [14]:
doc_examples = [
    "Course starts on 15th Jan 2024",
    "Prerequisites listed on GitHub",
    "Submit homeworks after start date",
    "Registration not required for participation",
    "Setup Google Cloud and Python before course"
]
cv_ex = CountVectorizer()
cv_ex.fit(doc_examples)

CountVectorizer()

In [15]:
X = cv_ex.transform(doc_examples)

In [16]:
pd.DataFrame(X.todense(), columns = cv_ex.get_feature_names_out()).T

0  1  2  3  4
15th           1  0  0  0  0
2024           1  0  0  0  0
after          0  0  1  0  0
and            0  0  0  0  1
before         0  0  0  0  1
cloud          0  0  0  0  1
course         1  0  0  0  1
date           0  0  1  0  0
for            0  0  0  1  0
github         0  1  0  0  0
google         0  0  0  0  1
homeworks      0  0  1  0  0
jan            1  0  0  0  0
listed         0  1  0  0  0
not            0  0  0  1  0
on             1  1  0  0  0
participation  0  0  0  1  0
prerequisites  0  1  0  0  0
python         0  0  0  0  1
registration   0  0  0  1  0
required       0  0  0  1  0
setup          0  0  0  0  1
start          0  0  1  0  0
starts         1  0  0  0  0
submit         0  0  1  0  0

We can see how many times each token is contained in each of the documents, being here the documents 0,1,2,3 and 4.
We can also disregard some words like "and", "for", "not" and other not relevant words which are named "stop words".

In [17]:
cv_ex = CountVectorizer(stop_words="english")
cv_ex.fit(doc_examples)

CountVectorizer(stop_words='english')

In [18]:
cv_ex.get_feature_names_out()

array(['15th', '2024', 'cloud', 'course', 'date', 'github', 'google',
       'homeworks', 'jan', 'listed', 'participation', 'prerequisites',
       'python', 'registration', 'required', 'setup', 'start', 'starts',
       'submit'], dtype=object)

In [19]:
X = cv_ex.transform(doc_examples)
pd.DataFrame(X.todense(), columns = cv_ex.get_feature_names_out()).T

0  1  2  3  4
15th           1  0  0  0  0
2024           1  0  0  0  0
cloud          0  0  0  0  1
course         1  0  0  0  1
date           0  0  1  0  0
github         0  1  0  0  0
google         0  0  0  0  1
homeworks      0  0  1  0  0
jan            1  0  0  0  0
listed         0  1  0  0  0
participation  0  0  0  1  0
prerequisites  0  1  0  0  0
python         0  0  0  0  1
registration   0  0  0  1  0
required       0  0  0  1  0
setup          0  0  0  0  1
start          0  0  1  0  0
starts         1  0  0  0  0
submit         0  0  1  0  0

In [20]:
# Summarizing the bag of words application
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer(min_df=5, stop_words='english')
X = cv.fit_transform(df.text)

names = cv.get_feature_names_out()
df_docs = pd.DataFrame(X.toarray(), columns=names).T
df_docs

0    1    2    3    4    5    6    7    8    9    ...  938  939  \
01          0    0    0    0    0    0    0    0    0    0  ...    0    0   
02          0    0    0    0    0    0    0    0    0    0  ...    0    0   
03          0    0    0    0    0    0    0    0    0    0  ...    0    0   
04          0    0    0    0    0    0    0    0    0    0  ...    0    0   
05          0    0    0    0    0    0    0    0    0    0  ...    0    0   
...       ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
yes         0    0    1    0    0    0    1    1    1    0  ...    0    0   
yml         0    0    0    0    0    0    0    0    0    0  ...    0    0   
youtube     0    0    0    0    0    0    0    0    0    1  ...    0    0   
zip         0    0    0    0    0    0    0    0    0    0  ...    0    0   
zoomcamp    0    1    0    0    0    2    0    0    0    0  ...    0    0   

          940  941  942  943  944  945  946  947  
01          0    0    0    0    0    0    0    0  
02          0    0    0    0    0    0    0    0  
03          0    0    0    0    0    0    0    0  
04          0    0    0    0    0    0    0    0  
05          0    0    0    0    0    0    0    0  
...       ...  ...  ...  ...  ...  ...  ...  ...  
yes         0    0    0    0    0    0    0    0  
yml         0    0    0    1    0    0    0    0  
youtube     0    0    0    0    0    0    0    0  
zip         0    0    0    0    0    0    0    0  
zoomcamp    0    0    0    0    0    0    0    1  

[1333 rows x 948 columns]

Another vectorizing stragey is TFIDF. Let's think of the word "yes" being contained in a lot of documents in contrast with "yml" contained in just a couple, We can think that when a word is more frequent it's more important than others. The TfidfVectorizer give more importance to more frequent terms.

In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer

cv = TfidfVectorizer(min_df=5, stop_words='english')
X = cv.fit_transform(df.text)

names = cv.get_feature_names_out()
df_docs = pd.DataFrame(X.toarray(), columns=names).T
df_docs.round(2)

0     1     2    3    4     5     6    7     8     9    ...  938  \
01        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
02        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
03        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
04        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
05        0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
...       ...   ...   ...  ...  ...   ...   ...  ...   ...   ...  ...  ...   
yes       0.0  0.00  0.28  0.0  0.0  0.00  0.21  0.2  0.15  0.00  ...  0.0   
yml       0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
youtube   0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.15  ...  0.0   
zip       0.0  0.00  0.00  0.0  0.0  0.00  0.00  0.0  0.00  0.00  ...  0.0   
zoomcamp  0.0  0.43  0.00  0.0  0.0  0.19  0.00  0.0  0.00  0.00  ...  0.0   

          939  940  941  942   943  944  945  946   947  
01        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
02        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
03        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
04        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
05        0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
...       ...  ...  ...  ...   ...  ...  ...  ...   ...  
yes       0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
yml       0.0  0.0  0.0  0.0  0.11  0.0  0.0  0.0  0.00  
youtube   0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
zip       0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.00  
zoomcamp  0.0  0.0  0.0  0.0  0.00  0.0  0.0  0.0  0.15  

[1333 rows x 948 columns]

In [22]:
query = "Do I need to know python to sign up for the January course?"

q = cv.transform([query])
q.toarray()

array([[0., 0., 0., ..., 0., 0., 0.]])

In [23]:
query_dict = dict(zip(names, q.toarray()[0]))
query_dict

{'01': 0.0,
 '02': 0.0,
 '03': 0.0,
 '04': 0.0,
 '05': 0.0,
 '06': 0.0,
 '09': 0.0,
 '10': 0.0,
 '100': 0.0,
 '11': 0.0,
 '12': 0.0,
 '127': 0.0,
 '13': 0.0,
 '14': 0.0,
 '15': 0.0,
 '16': 0.0,
 '17': 0.0,
 '19': 0.0,
 '1st': 0.0,
 '20': 0.0,
 '2019': 0.0,
 '2020': 0.0,
 '2021': 0.0,
 '2022': 0.0,
 '2023': 0.0,
 '2024': 0.0,
 '21': 0.0,
 '22': 0.0,
 '24': 0.0,
 '25': 0.0,
 '2pacx': 0.0,
 '30': 0.0,
 '35': 0.0,
 '403': 0.0,
 '42': 0.0,
 '50': 0.0,
 '5000': 0.0,
 '5431': 0.0,
 '5432': 0.0,
 '60': 0.0,
 '600': 0.0,
 '7077': 0.0,
 '80': 0.0,
 '8080': 0.0,
 '8888': 0.0,
 '9696': 0.0,
 'abhijit': 0.0,
 'able': 0.0,
 'abolade': 0.0,
 'absolute': 0.0,
 'accept': 0.0,
 'access': 0.0,
 'accordingly': 0.0,
 'account': 0.0,
 'accuracy': 0.0,
 'action': 0.0,
 'activate': 0.0,
 'actual': 0.0,
 'actually': 0.0,
 'add': 0.0,
 'added': 0.0,
 'adding': 0.0,
 'addition': 0.0,
 'additional': 0.0,
 'additionally': 0.0,
 'address': 0.0,
 'admin': 0.0,
 'advani': 0.0,
 'ahmed': 0.0,
 'ai': 0.0,
 'airflow': 0

In [24]:
doc_dict = dict(zip(names, X.toarray()[1]))
doc_dict

{'01': 0.0,
 '02': 0.0,
 '03': 0.0,
 '04': 0.0,
 '05': 0.0,
 '06': 0.0,
 '09': 0.0,
 '10': 0.0,
 '100': 0.0,
 '11': 0.0,
 '12': 0.0,
 '127': 0.0,
 '13': 0.0,
 '14': 0.0,
 '15': 0.0,
 '16': 0.0,
 '17': 0.0,
 '19': 0.0,
 '1st': 0.0,
 '20': 0.0,
 '2019': 0.0,
 '2020': 0.0,
 '2021': 0.0,
 '2022': 0.0,
 '2023': 0.0,
 '2024': 0.0,
 '21': 0.0,
 '22': 0.0,
 '24': 0.0,
 '25': 0.0,
 '2pacx': 0.0,
 '30': 0.0,
 '35': 0.0,
 '403': 0.0,
 '42': 0.0,
 '50': 0.0,
 '5000': 0.0,
 '5431': 0.0,
 '5432': 0.0,
 '60': 0.0,
 '600': 0.0,
 '7077': 0.0,
 '80': 0.0,
 '8080': 0.0,
 '8888': 0.0,
 '9696': 0.0,
 'abhijit': 0.0,
 'able': 0.0,
 'abolade': 0.0,
 'absolute': 0.0,
 'accept': 0.0,
 'access': 0.0,
 'accordingly': 0.0,
 'account': 0.0,
 'accuracy': 0.0,
 'action': 0.0,
 'activate': 0.0,
 'actual': 0.0,
 'actually': 0.0,
 'add': 0.0,
 'added': 0.0,
 'adding': 0.0,
 'addition': 0.0,
 'additional': 0.0,
 'additionally': 0.0,
 'address': 0.0,
 'admin': 0.0,
 'advani': 0.0,
 'ahmed': 0.0,
 'ai': 0.0,
 'airflow': 0

In [25]:
# We compute the simililarity using the dot product.
# The resuñt is a vector of similarity scores (cousine similarity)
X.dot(q.T).todense()

matrix([[0.19464486],
        [0.        ],
        [0.        ],
        [0.06011641],
        [0.04932915],
        [0.        ],
        [0.        ],
        [0.13477565],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.15899187],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.07431408],
        [0.        ],
        [0.        ],
        [0.05779673],
        [0.07243428],
        [0.        ],
        [0.05174293],
        [0.16373635],
        [0.08076031],
        [0.        ],
        [0.09755254],
        [0.        ],
        [0.21069625],
        [0.12067781],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.06381749],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.        ],
        [0.00910541],
        [0.02835681],
        [0.05480112],
        [0.        ],
        [0.        ],
        [0.        ],
        [0

In [26]:
from sklearn.metrics.pairwise import cosine_similarity

score = cosine_similarity(X, q).flatten()
score

array([0.19464486, 0.        , 0.        , 0.06011641, 0.04932915,
       0.        , 0.        , 0.13477565, 0.        , 0.        ,
       0.        , 0.15899187, 0.        , 0.        , 0.        ,
       0.07431408, 0.        , 0.        , 0.05779673, 0.07243428,
       0.        , 0.05174293, 0.16373635, 0.08076031, 0.        ,
       0.09755254, 0.        , 0.21069625, 0.12067781, 0.        ,
       0.        , 0.        , 0.        , 0.06381749, 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.00910541,
       0.02835681, 0.05480112, 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.02469964, 0.05129386, 0.06013439,
       0.05252658, 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.04169018, 0.        , 0.        , 0.        , 0.0075293 ,
       0.        , 0.        , 0.01971463, 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.     

In [27]:
import numpy as np

np.argsort(score)[-5:]

array([764,  27, 806, 577, 445], dtype=int64)

In [28]:
print(df.iloc[445].text)

Check this article. If you know everything in this article, you know enough. If you don’t, read the article and join the coursIntroduction to Pythone too :)
Introduction to Python – Machine Learning Bookcamp
You can follow this English course from the OpenClassrooms e-learning platform, which is free and covers the python basics for data analysis: Learn Python Basics for Data Analysis - OpenClassrooms . It is important to know some basics such as: how to run a Jupyter notebook, how to import libraries (and what libraries are), how to declare a variable (and what variables are) and some important operations regarding data analysis.
(Mélanie Fouesnard)


In [29]:
query2 = "I just found out about the course, is it too late to join"

q2 = cv.transform([query2])

s2 = cosine_similarity(X,q2).flatten()
np.argsort(s2)[-5:]

array([ 22, 448, 449, 440,   0], dtype=int64)

In [30]:
print(df.iloc[440].text)

The process is automated now, so you should receive the email eventually. If you haven’t, check your promotions tab in Gmail as well as spam.
If you unsubscribed from our newsletter, you won't get course related updates too.
But don't worry, it’s not a problem. To make sure you don’t miss anything, join the #course-ml-zoomcamp channel in Slack and our telegram channel with announcements. This is enough to follow the course.


In [31]:
fields = ["section", "question", "text"]
matrices = {}
vectorizers = {}

for f in fields:
    cv = TfidfVectorizer(stop_words="english", min_df=5)
    X = cv.fit_transform(df[f])
    matrices[f] = X
    vectorizers[f] = cv

In [32]:
matrices

{'section': <948x66 sparse matrix of type '<class 'numpy.float64'>'
 	with 3090 stored elements in Compressed Sparse Row format>,
 'question': <948x291 sparse matrix of type '<class 'numpy.float64'>'
 	with 3431 stored elements in Compressed Sparse Row format>,
 'text': <948x1333 sparse matrix of type '<class 'numpy.float64'>'
 	with 23808 stored elements in Compressed Sparse Row format>}

In [33]:
n = len(df)
n

948

In [34]:
score = np.zeros(n)

query = "I just discovered the course, is it too late to join?"

boosts = {
    "question": 3,
    #"text": 0.5
}

for f in fields:
    q = vectorizers[f].transform([query])
    X = matrices[f]

    f_scores = cosine_similarity(X, q).flatten()

    boost = boosts.get(f, 1.0)

    score = score + boost*f_scores

In [35]:
score

array([3.52985023, 3.49512426, 2.70735166, 2.96614194, 3.49512426,
       3.49512426, 1.93689291, 3.67069698, 2.67242848, 3.49512426,
       3.10198469, 2.46096752, 0.49512426, 0.49512426, 0.49512426,
       0.59193348, 0.49512426, 2.63772182, 0.57041627, 0.49512426,
       0.49512426, 0.49512426, 0.79499188, 0.60033101, 0.49512426,
       0.49512426, 0.49512426, 0.76959902, 0.62340833, 0.49512426,
       0.49512426, 0.49512426, 0.49512426, 1.78972334, 3.49512426,
       1.72080809, 0.49512426, 0.49512426, 0.49512426, 0.52668735,
       0.54427244, 2.00115141, 0.49512426, 0.53842198, 0.        ,
       0.        , 0.        , 0.        , 0.02804374, 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.06739038, 0.        , 0.00980845,
       0.        , 0.        , 0.        , 0.        , 0.05820102,
       0.        , 0.        , 0.        , 0.        , 0.     

In [36]:
filters = {
    "course": "data-engineering-zoomcamp"
}

for field, value in filters.items():
    mask = (df[field] == value).astype(int).values
    score = score*mask
score

array([3.52985023, 3.49512426, 2.70735166, 2.96614194, 3.49512426,
       3.49512426, 1.93689291, 3.67069698, 2.67242848, 3.49512426,
       3.10198469, 2.46096752, 0.49512426, 0.49512426, 0.49512426,
       0.59193348, 0.49512426, 2.63772182, 0.57041627, 0.49512426,
       0.49512426, 0.49512426, 0.79499188, 0.60033101, 0.49512426,
       0.49512426, 0.49512426, 0.76959902, 0.62340833, 0.49512426,
       0.49512426, 0.49512426, 0.49512426, 1.78972334, 3.49512426,
       1.72080809, 0.49512426, 0.49512426, 0.49512426, 0.52668735,
       0.54427244, 2.00115141, 0.49512426, 0.53842198, 0.        ,
       0.        , 0.        , 0.        , 0.02804374, 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.06739038, 0.        , 0.00980845,
       0.        , 0.        , 0.        , 0.        , 0.05820102,
       0.        , 0.        , 0.        , 0.        , 0.     

In [37]:
idx = np.argsort(-score)[:5]

In [38]:
df.iloc[idx]

course                           section  \
7  data-engineering-zoomcamp  General course-related questions   
0  data-engineering-zoomcamp  General course-related questions   
1  data-engineering-zoomcamp  General course-related questions   
4  data-engineering-zoomcamp  General course-related questions   
5  data-engineering-zoomcamp  General course-related questions   

                                            question  \
7  Course - Can I follow the course after it fini...   
0               Course - When will the course start?   
1  Course - What are the prerequisites for this c...   
4   Course - What can I do before the course starts?   
5             Course - how many Zoomcamps in a year?   

                                                text  
7  Yes, we will keep all the materials after the ...  
0  The purpose of this document is to capture fre...  
1  GitHub - DataTalksClub data-engineering-zoomca...  
4  You can start by installing and setting up all...  
5  There are 3 Zoom Camps in a year, as of 2024. ...

In [39]:
class TextSearch:

    def __init__(self, text_fields):
        self.text_fields = text_fields
        self.matrices = {}
        self.vectorizers = {}

    def fit(self, records, vectorizer_params={}):
        self.df = pd.DataFrame(records)

        for f in self.text_fields:
            cv = TfidfVectorizer(**vectorizer_params)
            X = cv.fit_transform(self.df[f])
            self.matrices[f] = X
            self.vectorizers[f] = cv

    def search(self, query, n_results=10, boost={}, filters={}):
        score = np.zeros(len(self.df))

        for f in self.text_fields:
            b = boost.get(f, 1.0)
            q = self.vectorizers[f].transform([query])
            s = cosine_similarity(self.matrices[f], q).flatten()
            score = score + b * s

        for field, value in filters.items():
            mask = (self.df[field] == value).values
            score = score * mask

        idx = np.argsort(-score)[:n_results]
        results = self.df.iloc[idx]
        return results.to_dict(orient='records')

In [77]:
index = TextSearch(
    text_fields=['section', 'question', 'text']
)
index.fit(documents)

index.search(
    query='I just singned up. Is it too late to join the course?',
    n_results=5,
    boost={'question': 3.0},
    filters={'course': 'data-engineering-zoomcamp'}
)

[{'text': "Yes, even if you don't register, you're still eligible to submit the homeworks.\nBe aware, however, that there will be deadlines for turning in the final projects. So don't leave everything for the last minute.",
  'section': 'General course-related questions',
  'question': 'Course - Can I still join the course after the start date?',
  'course': 'data-engineering-zoomcamp'},
 {'text': "The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
  'section': 'General course-related questions',
  'question': 'Course - When will the course start?',
  'course': 'data-engineerin

## Vector Search

It's useful when the words don't match exactly. 

A case is the singular value dimensionality, which reduces the dimensionality of the vectors while preserving as much information as possible. It's talked in the paper Latent Semantic Analysis (1990).

In [45]:
from sklearn.decomposition import TruncatedSVD
from sklearn.decomposition import PCA
X = matrices['text']
cv = vectorizers['text']

In [41]:
svd = TruncatedSVD(n_components=16)
X_emb = svd.fit_transform(X)

In [46]:
pca = PCA(n_components=16)
X_emb2 = pca.fit_transform(X)

In [43]:
# we reduced dimensionality to 61
X_emb.shape

(948, 16)

In [44]:
X_emb[0]

array([ 0.09653267, -0.08220561, -0.10238587, -0.07956506,  0.0689549 ,
       -0.05555886,  0.03116663, -0.13740608, -0.21514706,  0.2806271 ,
        0.12753578,  0.09501603, -0.10023097,  0.08480439,  0.02739981,
        0.04608201])

In [48]:
X_emb2[0]

array([-0.09217288, -0.03823908,  0.10488754,  0.10099538, -0.09716713,
        0.16278189, -0.21627752, -0.10040161,  0.05297374, -0.22559259,
        0.00987953, -0.11577717,  0.06429445,  0.00477542, -0.03468062,
       -0.01114715])

In [49]:
query = 'I just singned up. Is it too late to join the course?'

Q = cv.transform([query])
Q_emb = svd.transform(Q)
Q_emb[0]

array([ 0.05790381, -0.03851536, -0.05646039, -0.02814274,  0.04027229,
       -0.05972004,  0.01824095, -0.08818489, -0.14984109,  0.17840648,
        0.10095319,  0.08736904, -0.06516207,  0.0571745 ,  0.02983813,
        0.03280046])

In [51]:
np.dot(X_emb[0], Q_emb[0])

0.15274545645060916

In [50]:
Q_emb2 = pca.transform(Q)
Q_emb2[0]

array([-0.0636326 , -0.05328825,  0.03889679,  0.06304154, -0.09010477,
        0.11233692, -0.1544254 , -0.06440848,  0.02849726, -0.15439181,
        0.03239703, -0.0828833 ,  0.04549348,  0.02164793, -0.01865527,
        0.01539123])

In [52]:
np.dot(X_emb2[0], Q_emb2[0])

0.13501575943701444

In [53]:
scores = cosine_similarity(X_emb, Q_emb).flatten()
idx = np.argsort(-scores)[:10]
list(df.loc[idx].text)

["The purpose of this document is to capture frequently asked technical questions\nThe exact day and hour of the course will be 15th Jan 2024 at 17h00. The course will start with the first  “Office Hours'' live.1\nSubscribe to course public Google Calendar (it works from Desktop only).\nRegister before the course starts using this link.\nJoin the course Telegram channel with announcements.\nDon’t forget to register in DataTalks.Club's Slack and join the channel.",
 'Yes, you can. You won’t be able to submit some of the homeworks, but you can still take part in the course.\nIn order to get a certificate, you need to submit 2 out of 3 course projects and review 3 peers’ Projects by the deadline. It means that if you join the course at the end of November and manage to work on two projects, you will still be eligible for a certificate.',
 'If you have submitted two projects (and peer-reviewed at least 3 course-mates’ projects for each submission), you will get the certificate for the cour

In [54]:
scores = cosine_similarity(X_emb2, Q_emb2).flatten()
idx = np.argsort(-scores)[:10]
list(df.loc[idx].text)

['If you have submitted two projects (and peer-reviewed at least 3 course-mates’ projects for each submission), you will get the certificate for the course. According to the course coordinator, Alexey Grigorev, only two projects are needed to get the course certificate.\n(optional) David Odimegwu',
 'Yes, you can. You won’t be able to submit some of the homeworks, but you can still take part in the course.\nIn order to get a certificate, you need to submit 2 out of 3 course projects and review 3 peers’ Projects by the deadline. It means that if you join the course at the end of November and manage to work on two projects, you will still be eligible for a certificate.',
 'The course videos are pre-recorded, you can start watching the course right now.\nWe will also occasionally have office hours - live sessions where we will answer your questions. The office hours sessions are recorded too.\nYou can see the office hours as well as the pre-recorded course videos in the course playlist on

In [55]:
from sklearn.decomposition import NMF

nmf = NMF(n_components=16)
X_emb = nmf.fit_transform(X)
X_emb[0]

array([0.1282919 , 0.        , 0.        , 0.        , 0.        ,
       0.        , 0.00061603, 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 0.        ,
       0.        ])

In [57]:
Q = cv.transform([query])
Q_emb = nmf.transform(Q)
Q_emb[0]

scores = cosine_similarity(X_emb, Q_emb).flatten()
idx = np.argsort(-scores)[:10]
df.loc[idx]

course                           section  \
814             mlops-zoomcamp        +-General course questions   
2    data-engineering-zoomcamp  General course-related questions   
449  machine-learning-zoomcamp  General course-related questions   
11   data-engineering-zoomcamp  General course-related questions   
0    data-engineering-zoomcamp  General course-related questions   
764  machine-learning-zoomcamp   Projects (Midterm and Capstone)   
7    data-engineering-zoomcamp  General course-related questions   
451  machine-learning-zoomcamp  General course-related questions   
436  machine-learning-zoomcamp  General course-related questions   
450  machine-learning-zoomcamp  General course-related questions   

                                              question  \
814  What if my answer is not exactly the same as t...   
2    Course - Can I still join the course after the...   
449  The course has already started. Can I still jo...   
11   Certificate - Can I follow the course in a sel...   
0                 Course - When will the course start?   
764  What If I submitted only two projects and fail...   
7    Course - Can I follow the course after it fini...   
451      Can I submit the homework after the due date?   
436                      Is it going to be live? When?   
450                When does the next iteration start?   

                                                  text  
814  Please choose the closest one to your answer. ...  
2    Yes, even if you don't register, you're still ...  
449  Yes, you can. You won’t be able to submit some...  
11   No, you can only get a certificate if you fini...  
0    The purpose of this document is to capture fre...  
764  If you have submitted two projects (and peer-r...  
7    Yes, we will keep all the materials after the ...  
451  No, it’s not possible. The form is closed afte...  
436  The course videos are pre-recorded, you can st...  
450  The course is available in the self-paced mode...

## Creating embeddings with Bert



In [58]:
import torch

from transformers import BertModel, BertTokenizer

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertModel.from_pretrained("bert-base-uncased")
model.eval() # Set the model to evaluation mode if not training

c:\Users\User\Documents\LLM_zoomcamp\workshop\llm-zoomcamp-3.10\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
c:\Users\User\Documents\LLM_zoomcamp\workshop\llm-zoomcamp-3.10\lib\site-packages\huggingface_hub\file_download.py:157: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\User\.cache\huggingface\hub\models--bert-base-uncased. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Pyt

BertModel(
  (embeddings): BertEmbeddings(
    (word_embeddings): Embedding(30522, 768, padding_idx=0)
    (position_embeddings): Embedding(512, 768)
    (token_type_embeddings): Embedding(2, 768)
    (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
    (dropout): Dropout(p=0.1, inplace=False)
  )
  (encoder): BertEncoder(
    (layer): ModuleList(
      (0-11): 12 x BertLayer(
        (attention): BertAttention(
          (self): BertSdpaSelfAttention(
            (query): Linear(in_features=768, out_features=768, bias=True)
            (key): Linear(in_features=768, out_features=768, bias=True)
            (value): Linear(in_features=768, out_features=768, bias=True)
            (dropout): Dropout(p=0.1, inplace=False)
          )
          (output): BertSelfOutput(
            (dense): Linear(in_features=768, out_features=768, bias=True)
            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)
            (dropout): Dropout(p=0.1, inplace=False

In [59]:
texts = [
    "Yes, we will keep all the materials after the course finishes.",
    "You can follow the course at your own pace after it finishes"
]
encoded_input = tokenizer(texts, padding=True, truncation=True, return_tensors='pt')

In [60]:
encoded_input

{'input_ids': tensor([[  101,  2748,  1010,  2057,  2097,  2562,  2035,  1996,  4475,  2044,
          1996,  2607, 12321,  1012,   102],
        [  101,  2017,  2064,  3582,  1996,  2607,  2012,  2115,  2219,  6393,
          2044,  2009, 12321,   102,     0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0]])}

In [63]:
with torch.no_grad(): # Disable gradient calculation for inference
    outputs = model(**encoded_input)
    hidden_states = outputs.last_hidden_state

In [65]:
hidden_states.shape

torch.Size([2, 15, 768])

In [66]:
hidden_states[0]

tensor([[ 0.1010,  0.0181,  0.1303,  ..., -0.2932,  0.1863,  0.6615],
        [ 1.0608, -0.1242,  0.1370,  ..., -0.1605,  1.0429,  0.3532],
        [ 0.1802,  0.0776,  0.3941,  ..., -0.1379,  0.5974,  0.1704],
        ...,
        [ 0.4738, -0.0184,  0.2186,  ..., -0.0013, -0.0833, -0.2170],
        [ 0.6516,  0.1216, -0.2494,  ...,  0.1557, -0.5632, -0.4310],
        [ 0.7164,  0.2157, -0.0281,  ...,  0.2281, -0.6725, -0.3245]])

In [70]:
sentence_embeddings = hidden_states.mean(dim=1)
sentence_embeddings.shape

torch.Size([2, 768])

In [72]:
sentence_embeddings

tensor([[ 0.3600, -0.1607,  0.3545,  ...,  0.0429,  0.0348, -0.0382],
        [ 0.1785, -0.5000,  0.2528,  ..., -0.1141, -0.3361,  0.4110]])

In [73]:
X_emb = sentence_embeddings.numpy()

In [74]:
sentence_embeddings_cpu = sentence_embeddings.cpu()

In [75]:
def make_batches(seq, n):
    result = []
    for i in range(0, len(seq), n):
        batch = seq[i:i+n]
        result.append(batch)
    return result

In [78]:
from tqdm import tqdm 


texts = df['text'].tolist()
text_batches = make_batches(texts, 8)

all_embeddings = []

for batch in tqdm(text_batches):
    encoded_input = tokenizer(batch, padding=True, truncation=True, return_tensors='pt')

    with torch.no_grad():
        outputs = model(**encoded_input)
        hidden_states = outputs.last_hidden_state
        
        batch_embeddings = hidden_states.mean(dim=1)
        batch_embeddings_np = batch_embeddings.cpu().numpy()
        all_embeddings.append(batch_embeddings_np)

final_embeddings = np.vstack(all_embeddings)

  0%|          | 0/119 [00:00<?, ?it/s]

100%|██████████| 119/119 [11:09<00:00,  5.62s/it]


In [79]:
def compute_embeddings(texts, batch_size=8):
    text_batches = make_batches(texts, 8)
    
    all_embeddings = []
    
    for batch in tqdm(text_batches):
        encoded_input = tokenizer(batch, padding=True, truncation=True, return_tensors='pt')
    
        with torch.no_grad():
            outputs = model(**encoded_input)
            hidden_states = outputs.last_hidden_state
            
            batch_embeddings = hidden_states.mean(dim=1)
            batch_embeddings_np = batch_embeddings.cpu().numpy()
            all_embeddings.append(batch_embeddings_np)
    
    final_embeddings = np.vstack(all_embeddings)
    return final_embeddings

In [ ]:
X_text = compute_embeddings(df['text'].tolist())

In [90]:
X_q = compute_embeddings(df.question.tolist())

100%|██████████| 119/119 [00:37<00:00,  3.13it/s]


In [82]:
Q_text = compute_embeddings(["I just singned up. Is it too late to join the course?"])

100%|██████████| 1/1 [00:00<00:00, 14.93it/s]


In [84]:
Q_text.shape

(1, 768)

In [91]:
X_q.shape

(948, 768)

In [92]:
scores = cosine_similarity(X_q, Q_text).flatten()
idx = np.argsort(-scores)[:10]
df.loc[idx]

course                                   section  \
452  machine-learning-zoomcamp          General course-related questions   
449  machine-learning-zoomcamp          General course-related questions   
503  machine-learning-zoomcamp        2. Machine Learning for Regression   
448  machine-learning-zoomcamp          General course-related questions   
816             mlops-zoomcamp                +-General course questions   
3    data-engineering-zoomcamp          General course-related questions   
8    data-engineering-zoomcamp          General course-related questions   
454  machine-learning-zoomcamp          General course-related questions   
575  machine-learning-zoomcamp  4. Evaluation Metrics for Classification   
18   data-engineering-zoomcamp          General course-related questions   

                                              question  \
452  I just joined. What should I do next? How can ...   
449  The course has already started. Can I still jo...   
503  The answer I get for one of the homework quest...   
448  I’m new to Slack and can’t find the course cha...   
816  Can I still graduate when I didn’t complete ho...   
3    Course - I have registered for the Data Engine...   
8    Course - Can I get support if I take the cours...   
454  What’s the difference between the previous ite...   
575  I didn’t fully understand the ROC curve. Can I...   
18   Leaderboard - I am not on the leaderboard / ho...   

                                                  text  
452  Welcome to the course! Go to the course page (...  
449  Yes, you can. You won’t be able to submit some...  
503  That’s normal. We all have different environme...  
448  Here’s how you join a in Slack: https://slack....  
816  In order to obtain the certificate, completion...  
3    You don't need it. You're accepted. You can al...  
8    Yes, the slack channel remains open and you ca...  
454  There’s not much difference. There was one spe...  
575  It's a complex and abstract topic and it requi...  
18   When you set up your account you are automatic...

## Practical Implementation Aspects and Tools

Key concepts for fast retrieval are posting lists and inverted index, which allaw to index millions of documents and work fast.

Another interesting idea for vector search is LSH (Locally sensitive Hashing) particualrly the case of Random Projection. It basicalle takes a long embedding a tuns it into a binary vector and allows o take the cousine similarity between bits. (See lsh binning or Minning of Massive Datasets)